[Documentation] code walk traversal (#34)
TiarkRompf/lms-clean: doc/main/scala/lms/core/traversal.md

## Traverse

> The lms/core/traversal.scala define how one can traverse the LMS IR. Travering a tree-structured

Reviewer: ```suggestion
The `lms/core/traversal.scala` define how one can traverse the LMS IR. Travering a tree-structured
```

IR seems trivial, but traversing LMS IR (sea of the nodes) is not. The main challenge is to determine
the group of nodes that should be scoped in each block. Previously in [Backend](backend.md), we
talked about how the scope of the blocks are implicitly expressed via data dependencies and control

> dependencies. In traversal.scala, we will see how exactly we collect the nodes for each block using

Reviewer: ```suggestion
dependencies. In `traversal.scala`, we will see how exactly we collect the nodes for each block using
```

the dependency information.

The basic idea is as follows:
1. Each block has input-variables and input-effect-variables. We call them *bound variables*.
2. A node is ready to be schedule for the current block when the bound variables the node depends on are already in path (those blocks are enclosing the current block).

> 3. A node should be schedule in the current block (instead of in a inner block) when the node is used often.

Reviewer: ```suggestion
3. A node should be scheduled in the current block (instead of in a inner block) when the node is used often.
```

> 3. A node should be schedule in the current block (instead of in a inner block) when the node is used often.

Reviewer: ```suggestion
3. A node should be schedule in the current block (instead of in an inner block) when the node is used often.
```

### Dead Code Elimination

In a real scenario, we first run a dead code elimination (DCE) pass before we traverse the graph.

> The DCE pass is currently implemented in lms/core/backend.scala, `class DeadCodeElimCG`.

Reviewer: ```suggestion
The DCE pass is currently implemented in `class DeadCodeElimCG` (in file `lms/core/backend.scala`).
```

The basic idea for DCE is like this. We track 2 categories of `Backend.Sym`s, the *live* and *reach* sets.

> 1. The reach set is the set of nodes (or Syms of the nodes) that are reachable via hard-dependencies.

Reviewer: Maybe also explain a little bit hard-, data-, and control-dependencies?

Author: added in the paragraph above

It includes both data dependencies and control dependencies.

> 2. The live set is the set of nodes (or Syms of the nodes) that are needed only via data dependencies.

Reviewer: ```suggestion
2. The live set is the set of nodes (or `Backend.Sym`s of the nodes) that are needed only via data dependencies.
```
